### PR TITLE
Add flow types

### PR DESCRIPTION
**Pull request by Mxchaeltrxn**
Addresses #102

### CODE REVIEW

File: src/layers.js [hunk 1]
```diff
@@ -1,4 +1,25 @@
...
+        cy: number,
+        relative: boolean
+      }
+    | {
+        cx?: number,
+        cy?: number,
+        x: number,
+        y: number,
+        relative: boolean
+      },
```
Comment (haydn) [inline src/layers.js:20]: I don't think this type is quite right. You can pass `w` and `h` along with `x` and `y`, but they're always optional. Similarly, `cx`, `cy` and `relative` are all optional. I think it's more like this:

```js
{
  x?: number,
  y?: number,
  w?: number,
  h?: number,
  cx?: number,
  cy?: number,
  relative?: boolean
}
```

File: src/layers.js [hunk 2]
```diff
@@ -153,7 +184,10 @@ const transformBounds = (matrices, { x1, y1, x2, y2 }) => {
   };
 };
 
-const transformPoints = (matrices, points) =>
+const transformPoints = (
+  matrices: Array<TypeMatrix>,
+  points: Array<Array<number>>
+): Array<Array<number>> =>
```
Comment (haydn) [inline src/layers.js:189]: Flow let's you describe tuples, which would be a more accurate type for the points:

```js
Array<[number, number]>
```

File: src/layers.js [hunk 3]
```diff
@@ -164,7 +198,11 @@ const transformPoints = (matrices, points) =>
     )
   );
 
-const getExtremeBounds = (layers, extreme, predicate = () => true) => {
+const getExtremeBounds = (
+  layers: Array<TypeLayer>,
+  extreme: string,
```
Comment (haydn) [inline src/layers.js:202]:
```suggestion
  extreme: "widest" | "narrowest" | "tallest" | "shortest",
```

File: src/layers.js [hunk 4]
```diff
@@ -197,7 +235,11 @@ const getExtremeBounds = (layers, extreme, predicate = () => true) => {
   }
 };
 
-const resizeLayersToExtreme = (layers, extreme, predicate = () => true) => {
+const resizeLayersToExtreme = (
+  layers: Array<TypeLayer>,
+  extreme: string,
```
Comment (haydn) [inline src/layers.js:239]:
```suggestion
  extreme: "widest" | "narrowest" | "tallest" | "shortest",
```

File: src/types/types.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+export type TypeLayerPredicate = (layer: TypeLayer) => boolean;
+
+export type TypeMatrix = {
+  a: number,
+  b: number,
+  c: number,
+  d: number,
+  e: number,
+  f: number
+};
```
Comment (haydn) [inline src/types/types.js]: Let's leave off the `Type` prefix on the identifiers. We've not got much else using PascalCase except components, so it should be pretty easy to identify types where they're used and there's unlikely to be many naming conflicts.

File: src/layers.js [hunk 6]
```diff
@@ -146,8 +146,7 @@ const alignLayers = (
   );
 };
 
-// TODO: Rename this getSelectionBounds?
```
Comment (haydn) [inline src/layers.js:124]: Sorry! This comment was misleading. 😬 It would've made sense when I wrote it, but I don't think it makes sense anymore. The `getLayerBounds` function has no knowledge of the selection, so it's name is fine (I assume at the time I wrote this comment it when automatically detecting the selection or something like that).

Let's revert this commit.

File: src/types/types.js [hunk 7]
```diff
@@ -1,3 +1,4 @@
+// Layer transformation
```
Comment (haydn) [inline src/types/types.js:1]: What's this about?

File: src/types/types.js [hunk 8]
```diff
@@ -6,7 +7,7 @@ export type TypeLayer = {
   y1: number,
   x2: number,
   y2: number,
-  options: {}
+  options?: { [string]: string }
```
Comment (haydn) [inline src/types/types.js:10]: This type ain't right (the value can be all sorts of things). It'll be given a proper type in #86, but until then let's just give it an `any` type.

```js
options?: any
```

A useful trick for types like this is to setup a `TODO` type:

```js
type TODO = any
```

Which you can then use like this:

```js
options?: TODO
```

File: src/types/types.js (hunk 7 above)
Comment (Mxchaeltrxn) [reply src/types/types.js:1]: I added that to describe the types (I thought it might be useful when there are lots of types).

I'll just remove it for now and see what you say in response to this comment.

File: src/types/types.js (hunk 5 above)
Comment (Mxchaeltrxn) [reply src/types/types.js]: I have implemented this change but because I am curious, even if naming conflicts are unlikely to occur, isn't this a small price to pay to avoid any future problems in this area?

File: src/editor/exampleDoc.js [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import pushID from "../pushID";
```
[the remark this answers is not in the document]
Comment (Mxchaeltrxn) [reply src/editor/exampleDoc.js:1]: Out of curiosity, how did you notice this? (Asking because it's easy to miss).

Were you actively checking for // @flow at the top of each file, while reviewing (because you know this PR is for flow type)s?

**Comment by Mxchaeltrxn** [comment]
@haydn I've implemented the fixes you've mentioned and asked some questions. As always, thanks! 😄

Also, no one ever seems to comment on this but in my last 6 commits, I tried to add detail to the commit (adding a body for some commits) and make the message a bit clearer. Can you let me know what you think of that or if ... well, you think very detailed commit messages aren't actually that important.

File: src/editor/exampleDoc.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (haydn) [reply src/editor/exampleDoc.js:2]: I knew to look for it — I've been caught out by it many times before. 😀

There's an ESLint plugin that includes a check for the file annotation, but let's hold off on that until we're using Flow a bit more widely:

https://github.com/gajus/eslint-plugin-flowtype#require-valid-file-annotation

File: src/types/types.js (hunk 5 above)
Comment (haydn) [reply src/types/types.js]: I think it's a bit of an arbitrary distinction to call out types as needing a special prefix. We don't prefix every class with "ClassSomething" or every component with "ComponentSomething". That's because we're used to them and our tooling (editors, linters etc) can automatically identify them for us. Yet we rarely run into naming conflicts with those. So what problem are we solving by naming types is a special way?

If it was trivial to do, then not worries, but a convention like this is also pretty hard to follow consistently. Would interfaces by prefixed with "Type"? What about classes ([they can be used as types](https://flow.org/en/docs/types/classes/))? Pretty quickly you find yourself and in world where you're trying to guess if the type you want to used in prefixed with "Type" or not. 🙁

The other thing that typically happens is that you want to name a type "SomethingType"… a hypothetical example in this project would be a type called "Component" and another called "ComponentType" (the former describes an instance of a component, the latter is the component's definition as it is in the config file). If you're dogmatically prefixing everything with "Type" you'll end up with "TypeComponentType". 🙂

Anyway, long story short: naming things is hard and conventions are hard to enforce. I like to keep it simple and deal with the edge cases as they come up. Haha.

File: src/types/types.js [hunk 10]
```diff
@@ -6,7 +8,7 @@ export type Layer = {
   y1: number,
   x2: number,
   y2: number,
-  options?: { [string]: string }
+  options?: TODO
 };
 
 export type LayerPredicate = (layer: Layer) => boolean;
```
Comment (haydn) [inline src/types/types.js:14]: Not at al vital, but if you want you could make this a generic predicate type so we can reuse it for other types.

So instead of this:

```js
type LayerPredicate = (layer: Layer) => boolean;
const foo: LayerPredicate = layer => layer.id === id;
```

We have this:

```js
type Predicate<T> = (item: T) => boolean;
const foo: Predicate<Layer> = layer => layer.id === id;
const bar: Predicate<SketchbookDocument> = doc => doc.layers.length > 0;
```

We don't actually have any other cases where we're using a predicate like this yet, so you could argue it's premature abstraction to generalise it at this stage.

File: src/types/types.js [hunk 11]
```diff
@@ -1,3 +1,5 @@
+type TODO = any;
```
Comment (haydn) [inline src/types/types.js:1]: We're missing an `// @flow` here too! Which is probably why the `TypeLayer` usage below isn't throwing any errors. 🙂

File: src/types/types.js [hunk 12]
```diff
@@ -19,3 +19,8 @@ export type Matrix = {
   e: number,
   f: number
 };
+
+export type SketchbookDocument = {
+  type: "SketchbookDocument",
+  layers: Array<TypeLayer>
```
Comment (haydn) [inline src/types/types.js:27]: Got a `TypeLayer` here… which should be throwing an exception somewhere. 🤔 I wonder why the CI check didn't pick this up?

File: src/types/types.js (hunk 7 above)
Comment (haydn) [reply src/types/types.js]: Oh right. Yeah, for sure it can be useful to describe the types when there's a lot going on!

I think this particular case that comment could be misleading because the `Layer` type is (conceptually) used for more than just the transformations in `layers.js`. They appear in the document and get rendered by the canvas etc. You could actually argue that all the transformations in `layers.js` don't actually use the `Layer` type at all, but instead an inexact object with `x1`, `y1`, `x2` and `y2` properties:

```js
type TransformableObject = {
  x1: number,
  y1: number,
  x2: number,
  y2: number,
  ...
};
```

A layer just happens to conform to that type, but let's just leave it as is for now. 🙂

Comment (Mxchaeltrxn) [reply src/types/types.js]: Now that makes a lot of sense. I was unsure how to structure the type so that it must have certain fields but didn't have to have other fields. It was erroring because layer was being passed in but my type (originally) was just 
```
{
  x1: number,
  y1: number,
  x2: number,
  y2: number,
}
```

File: src/types/types.js (hunk 12 above)
Comment (Mxchaeltrxn) [reply src/types/types.js:27]: Ah woops. I didn't realise I had to put `@flow` in the type file. Fixed!

**Comment by Mxchaeltrxn** [comment]
Fixed! I've unresolved some of your comments but that's just because I want to be able to see those educational nuggets without opening them. I'll write them down later.

**Comment by haydn** [comment]
LGTM!
